**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №9

Набор данных: auto.csv

Атрибуты:
...

### 1.Чтение файла (набора данных)

In [24]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd

df = pd.read_csv('auto.csv', sep=',')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [25]:
# применить метод head
df.head(20)

Name  year  SellingPrice     kmdriven     fuel  \
0                     Maruti 800 AC  2007   60000.00000  70000.00000   Petrol   
1          Maruti Wagon R LXI Minor  2007  135000.00000  50000.00000   Petrol   
2              Hyundai Verna 1.6 SX  2012  600000.00000 100000.00000   Diesel   
3            Datsun RediGO T Option  2017  250000.00000  46000.00000   Petrol   
4             Honda Amaze VX i-DTEC  2014  450000.00000 141000.00000   Diesel   
5              Maruti Alto LX BSIII  2007  140000.00000 125000.00000   Petrol   
6         Hyundai Xcent 1.2 Kappa S  2016  550000.00000  25000.00000   Petrol   
7         Tata Indigo Grand Petroll  2014  240000.00000  60000.00000   Petrol   
8          Hyundai Creta 1.6 VTVT S  2015  850000.00000  25000.00000  Petroll   
9          Maruti Celerio Green VXI  2017  365000.00000  78000.00000      CNG   
10          Chevrolet Sail 1.2 Base  2015  260000.00000  35000.00000   Petrol   
11         Tata Indigo Grand Petrol  2014  250000.00000 100000.00000   Petrol   
12  Toyota Corolla Altis 1.8 VL CVT  2018 1650000.00000  25000.00000   Petrol   
13                    Maruti 800 AC  2007   60000.00000  70000.00000   Petrol   
14         Maruti Wagon R LXI Minor  2007  135000.00000  50000.00000   Petrol   
15             Hyundai Verna 1.6 SX  2012  600000.00000 100000.00000   Diesel   
16           Datsun RediGO T Option  2017  250000.00000  46000.00000   Petrol   
17            Honda Amaze VX i-DTEC  2014  450000.00000 141000.00000   Diesel   
18             Maruti Alto LX BSIII  2007  140000.00000 125000.00000   Petrol   
19        Hyundai Xcent 1.2 Kappa S  2016  550000.00000  25000.00000   Petrol   

   seller_Type transmission         owner  
0   Individual       Manual   First Owner  
1   Individual       Manual   First Owner  
2   Individual       Manual   First Owner  
3   Individual       Manual   First Owner  
4   Individual       Manual  Second Owner  
5   Individual       Manual   First Owner  
6   Individual       Manual   First Owner  
7   Individual       Manual  Second Owner  
8   Individual       Manual   First Owner  
9   Individual       Manual   First Owner  
10  Individual       Manual   First Owner  
11  Individual       Manual   First Owner  
12      Dealer    Automatic   First Owner  
13  Individual       Manual   First Owner  
14  Individual       Manual   First Owner  
15  Individual       Manual   First Owner  
16  Individual       Manual   First Owner  
17  Individual       Manual  Second Owner  
18  Individual       Manual   First Owner  
19  Individual       Manual   First Owner

2.2 Оценка данных с помощью метода info.

In [26]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4344 entries, 0 to 4343
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          4344 non-null   object 
 1   year          4344 non-null   int64  
 2   SellingPrice  4342 non-null   float64
 3   kmdriven      4343 non-null   float64
 4   fuel          4343 non-null   object 
 5   seller_Type   4344 non-null   object 
 6   transmission  4344 non-null   object 
 7   owner         4344 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 271.6+ KB


2.3 Оценка данных с помощью метода describe.

In [27]:
# оцените числовые столбцы с помощью describe
pd.set_option('display.float_format', '{:.5f}'.format)
df.describe()

year  SellingPrice     kmdriven
count 4344.00000    4342.00000   4343.00000
mean  2013.09231  503998.74090  66193.06332
std      4.21454  578446.45972  46636.64876
min   1992.00000   20000.00000      1.00000
25%   2011.00000  210000.00000  35000.00000
50%   2014.00000  350000.00000  60000.00000
75%   2016.00000  600000.00000  90000.00000
max   2020.00000 8900000.00000 806599.00000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**

 ---
 ## Промежуточный вывод о данных
Данные содержат 4344 записей о поддержанных авто с 8 признаками, включая имя владельца, год выпуска машины, цену продажи, пробег, тип топлива, тип продавца, тип коробки передач и тип владельца.

Типы данных: Числовые: year, SellingPrice, kmdriven.

Категориальные: Name, fuel, seller_Type, transmission, owner.

Пропуски: SellingPrice - 2 пропущенных значения, kmdriven и fuel — по 1 пропущенному значению.


 2.4 Оценка названий столбцов

In [28]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['Name', 'year', 'SellingPrice', 'kmdriven', 'fuel', 'seller_Type',
       'transmission', 'owner'],
      dtype='object')

In [29]:
# Проверка на наличие возможных проблем
problems_found = []

for col in df.columns:
    if col != col.strip():
        problems_found.append(f"Столбец '{col}' имеет пробелы в начале или конце")

for col in df.columns:
    if any(char in col for char in [' ', '.', '-', '/', '\\', '(', ')', '[', ']', '{', '}']):
        problems_found.append(f"Столбец '{col}' содержит специальные символы или пробелы")

if problems_found:
    print("Обнаружены проблемы в названиях столбцов:")
    for problem in problems_found:
        print(f"- {problem}")
else:
    print("Проблем не обнаружено")

Проблем не обнаружено


## Анализ и коррекция названий столбцов
Был проведен анализ названий столбцов на предмет наличия проблем:

- Проверка на пробелы в начале/конце названий
- Проверка на наличие специальных символов и пробелов

После анализа выполнено переименование столбцов для улучшения читаемости и удобства работы:

- Name → car_name
- year → release_year
- SellingPrice → selling_price
- kmdriven → km_driven
- fuel → fuel_type
- seller_Type → seller_type
- transmission → transmission
- owner → owner_type

В конце выводится обновленный список названий столбцов для подтверждения изменений.

In [30]:
# переименование столбцов
df.rename(columns={
    'Name': 'car_name',
    'year': 'release_year',
    'SellingPrice': 'selling_price',
    'kmdriven': 'km_driven',
    'fuel': 'fuel_type',
    'seller_Type': 'seller_type',
    'transmission': 'transmission',
    'owner': 'owner_type'
}, inplace=True)

print("Обновленные названия столбцов:")
print(df.columns.tolist())

Обновленные названия столбцов:
['car_name', 'release_year', 'selling_price', 'km_driven', 'fuel_type', 'seller_type', 'transmission', 'owner_type']


### 3. Проверка пропусков

In [31]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
import numpy as np

# Проверка пропусков
missing_data = df.isna().sum()
print(missing_data[missing_data > 0])

if missing_data.any():
    # Обработка числовых столбцов
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    for col in numeric_cols:
        if missing_data[col] > 0:
            median_val = df[col].median()
            df[col] = df[col].fillna(median_val)
            print(f"Заполнено {missing_data[col]} пропусков в числовом столбце '{col}' медианой: {median_val}")
    
    # Обработка остальных столбцов
    other_cols = df.columns.difference(numeric_cols)
    for col in other_cols:
        if missing_data[col] > 0:
            # Для строковых/категориальных данных используем моду
            if df[col].dtype == 'object':
                mode_val = df[col].mode()[0] if not df[col].mode().empty else 'Unknown'
                df[col] = df[col].fillna(mode_val)
                print(f"Заполнено {missing_data[col]} пропусков в столбце '{col}' модой: '{mode_val}'")
            
    remaining_missing = df.isna().sum().sum()
    if remaining_missing == 0:
        print("Все пропуски успешно обработаны!")
    else:
        print(f"\nОсталось {remaining_missing} пропусков, требующих дополнительного анализа.")
        print("Столбцы с оставшимися пропусками:")
        print(df.isnull().sum()[df.isna().sum() > 0])
else:
    print("Пропуски в данных отсутствуют.")

selling_price    2
km_driven        1
fuel_type        1
dtype: int64
Заполнено 2 пропусков в числовом столбце 'selling_price' медианой: 350000.0
Заполнено 1 пропусков в числовом столбце 'km_driven' медианой: 60000.0
Заполнено 1 пропусков в столбце 'fuel_type' модой: 'Diesel'
Все пропуски успешно обработаны!


In [32]:
missing_data = df.isna().sum()
print("Пропуски в данных:")
print(missing_data[missing_data > 0])

Пропуски в данных:
Series([], dtype: int64)


### Выбранные стратегии обработки пропусков:

1. Заполнение числовых данных медианой:

- Обоснование: Медиана более устойчива к выбросам, чем среднее значение
- Преимущество: Сохраняет общее распределение данных без смещения в сторону экстремальных значений

2. Заполнение категориальных данных модой:

- Обоснование: Мода представляет наиболее часто встречающееся значение
- Преимущество: Сохраняет исходное распределение категориальных переменных
### Альтернативные методы
1. Удаление строк с пропусками:

- Недостаток: Потеря данных и возможное смещение выборки
- Когда уместно: Только при малом количестве пропусков (<1%)

2. Создание отдельной категории "Unknown":

- Применимость: Для категориальных данных
- Преимущество: Сохранение информации о наличии пропуска

### Критерии выбора метода:
- Процент пропущенных значений (<5% - заполнение, >30% - удаление столбца)
- Тип данных (числовой/категориальный/временной)
- Важность признака для дальнейшего анализа
- Сохранение распределения и статистических свойств данных


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [33]:
# Проверка на наличие явных дубликатов
duplicates_count = df.duplicated().sum()
print(f"Количество полных дубликатов в данных: {duplicates_count}")

Количество полных дубликатов в данных: 763


In [34]:
# удалите дубликаты, если они есть
if duplicates_count > 0:
    df = df.drop_duplicates()
    print(f"Удалено {duplicates_count} дубликатов.")

Удалено 763 дубликатов.


#### Проверка неявных дубликатов

In [35]:
# Проверка на наличие неявных дубликатов
categorical_columns = df.select_dtypes(include=['object']).columns.tolist()

print("Категориальные столбцы для проверки неявных дубликатов:")
print(categorical_columns)

# Создаем словарь для хранения уникальных значений каждого категориального столбца
unique_values = {}

# Анализируем каждый категориальный столбец
for column in categorical_columns:
    print(f"\n Анализ столбца: {column}")
    print("Уникальные значения и их частота:")
    # Получаем частоту каждого значения
    value_counts = df[column].value_counts()
    print(value_counts)
    
    # Сохраняем уникальные значения для дальнейшего анализа
    unique_values[column] = value_counts.index.tolist()
    

Категориальные столбцы для проверки неявных дубликатов:
['car_name', 'fuel_type', 'seller_type', 'transmission', 'owner_type']

 Анализ столбца: car_name
Уникальные значения и их частота:
car_name
Maruti Swift Dzire VDI                            54
Maruti Alto 800 LXI                               48
Maruti Alto LXi                                   42
Maruti Alto LX                                    30
Maruti Swift VDI BSIV                             28
                                                  ..
Mahindra Verito Vibe 1.5 dCi D6                    1
Ford EcoSport 1.5 Petrol Titanium Plus AT BSIV     1
Tata Harrier XE                                    1
Hyundai Xcent 1.2 CRDi SX                          1
Fiat Linea Emotion (Diesel)                        1
Name: count, Length: 1492, dtype: int64

 Анализ столбца: fuel_type
Уникальные значения и их частота:
fuel_type
Diesel      1801
Petrol      1719
CNG           37
LPG           22
Petroll        1
Electric       1
Name: 

In [ ]:
# Создаем копию DataFrame для безопасного изменения
df_clean = df.copy()

# Исправление неявных дубликатов в столбце fuel_type
print("Исправление неявных дубликатов в столбце fuel_type:")

# Заменяем 'Petroll' на 'Petrol' с использованием replace
df_clean['fuel_type'] = df['fuel_type'].replace('Petroll', 'Petrol')
print("Исправлена опечатка: с 'Petroll' на 'Petrol'")

# Проверяем результат
print("\nОбновленное распределение значений в fuel_type:")
print(df_clean['fuel_type'].value_counts())

# Исправление неявных дубликатов в столбце seller_type
print("\nИсправление неявных дубликатов в столбце seller_type:")

# Заменяем 'Dealeer' на 'Dealer' с использованием replace
df_clean['seller_type'] = df['seller_type'].replace('Dealeer', 'Dealer')
print("Исправлена опечатка: с 'Dealeer' на 'Dealer'")

# Проверяем результат
print("\nОбновленное распределение значений в seller_type:")
print(df_clean['seller_type'].value_counts())

# Теперь присваиваем очищенный DataFrame обратно в df
df = df_clean

Исправление неявных дубликатов в столбце fuel_type:
Исправлена опечатка: с 'Petroll' на 'Petrol'

Обновленное распределение значений в fuel_type:
fuel_type
Diesel      1801
Petrol      1720
CNG           37
LPG           22
Electric       1
Name: count, dtype: int64

Исправление неявных дубликатов в столбце seller_type:
Исправлена опечатка: с 'Dealeer' на 'Dealer'

Обновленное распределение значений в seller_type:
seller_type
Individual          2836
Dealer               712
Trustmark Dealer      33
Name: count, dtype: int64


In [ ]:
# удалите дубликаты, если они есть
subset_columns = ['car_name', 'release_year', 'selling_price', 'km_driven', 'fuel_type', 'transmission',]
implicit_duplicates = df.duplicated(subset=subset_columns, keep=False)

# Количество неявных дубликатов
duplicate_count = implicit_duplicates.sum()
print(f"Количество неявных дубликатов: {duplicate_count}")

# Примеры неявных дубликатов (если они есть)
if duplicate_count > 0:
    print("\nПримеры неявных дубликатов:")
    print(df[implicit_duplicates].sort_values(by=subset_columns).head(duplicate_count))
df = df.drop_duplicates(subset=subset_columns, ignore_index=True)
print(df.info())


Количество неявных дубликатов: 10

Примеры неявных дубликатов:
                            car_name  release_year  selling_price  \
8           Hyundai Creta 1.6 VTVT S          2015   850000.00000   
21          Hyundai Creta 1.6 VTVT S          2015   850000.00000   
487   Mahindra Bolero DI DX 7 Seater          2007   225000.00000   
1315  Mahindra Bolero DI DX 7 Seater          2007   225000.00000   
1364                 Maruti Alto LXi          2007   100000.00000   
3804                 Maruti Alto LXi          2007   100000.00000   
1969                 Maruti Ritz VDi          2011   200000.00000   
4291                 Maruti Ritz VDi          2011   200000.00000   
2468        Maruti Wagon R VXI BS IV          2014   210000.00000   
2469        Maruti Wagon R VXI BS IV          2014   210000.00000   

        km_driven fuel_type seller_type transmission    owner_type  
8     25000.00000    Petrol  Individual       Manual   First Owner  
21    25000.00000    Petrol  Individual

### Выбор ключевых признаков:

Код использует подмножество столбцов: ['car_name', 'release_year', 'selling_price', 'km_driven', 'fuel_type', 'transmission']

Эти поля выбраны как комбинация, которая с высокой вероятностью должна быть уникальной для каждой машины
Поиск дубликатов:

- Метод df.duplicated(subset=subset_columns, keep=False) находит все строки, где комбинация указанных столбцов совпадает
- Параметр keep=False помечает все дубликаты (а не только вторые и последующие вхождения)
#### Анализ результатов:

- Подсчитывается количество неявных дубликатов
- При наличии дубликатов выводятся примеры для визуальной проверки

Также стоит омтетить, что записи не являются полными копиями (различаются в других столбцах), но совпадают по ключевым идентифицирующим признакам

---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [46]:
df.dtypes

car_name          object
release_year       int64
selling_price    float64
km_driven        float64
fuel_type         object
seller_type       object
transmission      object
owner_type        object
dtype: object

In [50]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
# Проверяем целочисленность и конвертируем столбцы с количественными данными
for col in ['selling_price', 'km_driven']:
    # Используем .loc для явного изменения столбцов
    if (df[col] % 1 == 0).all():
        df.loc[:, col] = df[col].astype('int64')

# Список категориальных столбцов
categorical_columns = ['fuel_type', 'seller_type', 'transmission', 'owner_type']
# Конвертируем категориальные столбцы
for col in categorical_columns:
    # Используем .loc для явного изменения столбцов
    df.loc[:, col] = df[col].astype('category')

print("Итоговые типы данных:")
print(df.dtypes)


Итоговые типы данных:
car_name           object
release_year        int64
selling_price       int64
km_driven           int64
fuel_type        category
seller_type      category
transmission     category
owner_type       category
dtype: object


Столбцы selling_price и km_driven указаны как float64, но при просмотре данных видно, что цены указаны только в рублях, без копеек а пройденные километры также только в целочисленном виде, эти столбцы были преобразованы в тип int64. Для категориальных признаков fuel_type, seller_type, transmission и owner_type текущий тип object не оптимален: данные в этих столбцах содержат ограниченный набор значений (например, Manual, Automatic), поэтому их необходимо перевести в тип category для экономии памяти и ускорения операций группировки.

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Группировка - тип коробки передач (transmission) по количеству
разных типов топлива (fuel)`*

In [60]:
# выполните группировку согласно варианту
transmission_fuel_counts = df.groupby('transmission')['fuel_type'].value_counts()
print(transmission_fuel_counts)

transmission  fuel_type
Automatic     Diesel        170
              Petrol        141
              Electric        1
              CNG             0
              LPG             0
Manual        Diesel       1629
              Petrol       1576
              CNG            37
              LPG            22
              Electric        0
Name: count, dtype: int64


C:\Users\fedor\AppData\Local\Temp\ipykernel_28640\3234380938.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  transmission_fuel_counts = df.groupby('transmission')['fuel_type'].value_counts()


**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

В результате группировки данных видно, что в изначальном датасете сильно преобладают машины с механической коробкой передач. Электромобили крайне редки и представлены только с коробкой-автомат на механике таких машин нет. В свою очередь на автоматической коробке передач нет машин, которые едут на газу (типы LPG и CNG). Большинство машин едут на дизеле и бензине, пропорция машин на таких видах топлива довольно схожа, что на автоматической, что на механической коробках передач, но все же на автоматических заметно небольшое преобладание дизельных автомобилей.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Группировка - тип владельца (owner) по количеству разных коробок
передач (transmission). Создать датафрейм. Переименовать столбец с
количеством в “сount”. Отсортировать по убыванию столбца “count”`*

In [67]:
# выполните группировку согласно варианту
# Группировка по 'owner' и подсчет 'transmission'
owner_transmission_counts = df.groupby('owner_type')['transmission'].value_counts().reset_index(name='count')


# Сортировка по убыванию столбца 'count'
owner_transmission_counts_sorted = owner_transmission_counts.sort_values(by='count', ascending=False)
print(owner_transmission_counts_sorted)

             owner_type transmission  count
0           First Owner       Manual   1991
4          Second Owner       Manual    914
8           Third Owner       Manual    270
1           First Owner    Automatic    227
2  Fourth & Above Owner       Manual     73
5          Second Owner    Automatic     64
9           Third Owner    Automatic     18
6        Test Drive Car       Manual     16
3  Fourth & Above Owner    Automatic      2
7        Test Drive Car    Automatic      1


C:\Users\fedor\AppData\Local\Temp\ipykernel_28640\1592056404.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  owner_transmission_counts = df.groupby('owner_type')['transmission'].value_counts().reset_index(name='count')


**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Большинство автомобилей в этой выборке имеют механическую коробку передач, особенно среди первых и вторых владельцев. Три верхние позиции по количеству занимают автомобили с механической коробкой. Количество автомобилей значительно уменьшается по мере увеличения числа предыдущих владельцев. Это ожидаемо, так как чем больше владельцев, тем меньше таких автомобилей на рынке. Также отдельно видно, что машин на тест-драйв очень мало и абсолютное большинство из них с автоматической коробкой передач.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Сводная таблица (pivot_table) - средняя стоимость для каждого года
year. Отсортировать по убыванию. Округлить до трёх знаков.`*

In [75]:
# выполните сводную таблицу согласно варианту
data_pivot = df.pivot_table(index=['release_year'], values='selling_price', aggfunc='mean')
data_pivot = data_pivot.sort_values(by='selling_price', ascending=False)
data_pivot['selling_price'] = data_pivot['selling_price'].round(3)
print(data_pivot)

              selling_price
release_year               
2019             865364.167
2018             820420.305
2020             804844.400
2017             740057.594
2016             638342.707
2014             512943.879
2015             510051.160
2013             433046.285
2012             379352.027
2011             296418.650
2010             275936.886
2009             235467.420
1996             225000.000
2008             177668.171
2007             165330.426
1998             165111.111
2006             154182.763
2004             126697.342
2005             114821.443
2001             109562.438
1995              95000.000
1997              93000.000
2003              87090.909
2002              86111.111
2000              81500.000
1999              73888.889
1992              50000.000


**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

В этих данных наблюдается четкая тенденция к росту цен с уменьшением возраста. Чем новее год выпуска, тем выше средняя цена продажи. Все данные можно разбить на 3 периода: новые, средние и старые.

Годы 2018–2020 имеют среднюю цену выше 800000. Годы 2009–2017 находятся в диапазоне от 235000 до 740000. Годы до 2009 (за исключением 1996) имеют среднюю цену значительно ниже 200000.

Заметен значительный разрыв между новыми и старыми машинами: существует резкий переход в цене. Машины, выпущенные после 2009 года, в среднем значительно дороже тех, что были выпущены до 2009 года.

Также видна аномалия 1996 года, он выделяется из общего тренда. Его средняя цена (225000) необычно высока для своего возраста и даже превосходит многие более новые годы. Ожидаемо, самые низкие средние цены наблюдаются у товаров, выпущенных в девяностых и начале двухтысячных годов.


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Сводная таблица (pivot_table) - среднее цена для каждого типа
владельца - столбцы и типа продавца- строки. Отсортировать по возрастанию
seller_type. Округлить до двух знаков.`*

In [78]:
# выполните сводную таблицу согласно варианту
pivot_table = df.pivot_table(index=['seller_type'], columns='owner_type', values='selling_price', aggfunc='mean')
pivot_table = pivot_table.sort_index(ascending=True)
pivot_table = pivot_table.round(2)
print(pivot_table)

owner_type        First Owner  Fourth & Above Owner  Second Owner  \
seller_type                                                         
Dealer              679505.08             117499.50     468499.94   
Individual          518769.48             182958.88     326973.98   
Trustmark Dealer    823709.68                   NaN     800000.00   

owner_type        Test Drive Car  Third Owner  
seller_type                                    
Dealer                 954293.94    462333.22  
Individual                   NaN    260014.69  
Trustmark Dealer             NaN          NaN  


C:\Users\fedor\AppData\Local\Temp\ipykernel_28640\2159261277.py:2: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table = df.pivot_table(index=['seller_type'], columns='owner_type', values='selling_price', aggfunc='mean')


**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

Автомобили от первого владельца, как правило, имеют самую высокую среднюю стоимость продажи среди всех категорий. Это логично, так как такие автомобили обычно новее и находятся в лучшем состоянии. 

Автомобили, у которых было четыре или более владельцев, имеют значительно более низкую среднюю стоимость продажи. Это ожидаемо, так как такие автомобили, вероятно, старше и/или имеют больший износ. Средняя цена для тестовых автомобилей, продаваемых дилерами, очень высока (954293). Это может указывать на то, что это были почти новые модели или модели высокого класса.

 В среднем самые высокие цены у Trustmark дилеров, далее идут обычные дилеры, у которых тоже дорогие машины и затем индивидуальные продавцы, у которых самые низкие цены. 
 
 Видно, что trustmark дилеры продают только автомобили с одним или двумя владельцами, а тест-драйв машины продают только обычные дилеры.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

 В ходе выполнения лабораторной работы был проведен предварительный анализ и предобработка набора данных auto.csv, содержащего информацию о поддержанных автомобилях, выставленных на продажу.

#### На этапе предобработки данных были выполнены следующие шаги:
Стандартизация названий столбцов: все названия были приведены к единому стилю. Обработка пропусков: Пропущенные значения в числовых столбцах были заменены медианными значениями, а в категориальных данных были заменены модой. Удаление дубликатов: были устранены как полные, так и неявные дубликаты. Была проведена оптимизация типов данных
#### Анализ данных с помощью группировок и сводных таблиц позволил сделать следующие выводы:
-	На рынке преобладают автомобили с механической коробкой передач, особенно среди машин с одним или двумя владельцами. Автомобили на газу (LPG/CNG) представлены только с механической коробкой передач, в то время как немногочисленные электромобили — только с автоматической.
-	Существует прямая зависимость между годом выпуска автомобиля и его средней ценой: чем новее автомобиль, тем он дороже. Наблюдается заметный скачок цен для автомобилей, выпущенных после 2009 года (не считая аномалию 1996 года). 
-	Количество предыдущих владельцев и тип продавца значительно влияют на стоимость. Наиболее дорогие автомобили продаются дилерами категории "Trustmark" и имеют не более двух владельцев. Самые низкие цены у частных продавцов.



### Дополнительное задание

**`Подробная формулировка задания`**

In [ ]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***